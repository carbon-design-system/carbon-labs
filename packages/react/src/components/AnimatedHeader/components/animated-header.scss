/**
 * Copyright IBM Corp. 2025
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

@use 'AnimatedHeader/animated-header.scss';
@use 'HeaderTitle/header-title.scss';
@use 'Tiles/AITile/ai-tile.scss';
@use 'Tiles/AIPromptTile/ai-prompt-tile.scss';
@use 'Tiles/GlassTile/glass-tile.scss';
