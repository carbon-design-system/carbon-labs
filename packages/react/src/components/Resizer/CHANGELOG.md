# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [0.2.0](https://github.com/carbon-design-system/carbon-labs/compare/@carbon-labs/react-resizer@0.1.0...@carbon-labs/react-resizer@0.2.0) (2025-06-24)


### Bug Fixes

* **Resizer:** update entry points for package ([#680](https://github.com/carbon-design-system/carbon-labs/issues/680)) ([f92ffa8](https://github.com/carbon-design-system/carbon-labs/commit/f92ffa8c21566b8bed5000605631d61079756f24))





# 0.1.0 (2025-06-12)


### Features

* Resizer initial commit ([#579](https://github.com/carbon-design-system/carbon-labs/issues/579)) ([d085ab5](https://github.com/carbon-design-system/carbon-labs/commit/d085ab5e6f6a562d0f9220eea0d25b6404ed7d4a))
