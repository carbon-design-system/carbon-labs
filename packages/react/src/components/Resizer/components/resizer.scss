/**
 * Copyright IBM Corp. 2025
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

// Carbon setting imports
@use '@carbon/styles/scss/theme' as *;
@use '@carbon/styles/scss/colors' as *;
@use '@carbon/styles/scss/spacing' as *;
@use '@carbon/styles/scss/components/button' as *;

$prefix: 'clabs' !default;

.#{$prefix}--resizer {
  position: relative;
  flex: none;
  background-color: $border-subtle-01;

  &:hover {
    @media (prefers-reduced-motion: no-preference) {
      background-color: $border-interactive;
      transition: background-color 150ms ease;
    }
  }

  &:focus {
    background-color: $border-interactive;
    outline: none;
  }

  &:active {
    background-color: $border-interactive;
  }

  &:focus:not(:focus-visible) {
    box-shadow: none;
    outline: none;
  }

  &--horizontal {
    block-size: 0.3rem;
    cursor: ns-resize;
  }

  &--vertical {
    inline-size: 0.3rem;
    cursor: ew-resize;
  }
}

.sr-only {
  position: absolute;
  inline-size: 1px;
  block-size: 1px;
  padding: 0;
  margin: -1px;
  overflow: hidden;
  clip: rect(0, 0, 0, 0);
  white-space: nowrap;
  border: 0;
}

// add transitions to panel resizes
.smooth-resize {
  @media (prefers-reduced-motion: no-preference) {
    transition: all 150ms linear;
  }
}
