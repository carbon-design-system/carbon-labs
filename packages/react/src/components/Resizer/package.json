{
  "name": "@carbon-labs/react-resizer",
  "version": "0.2.0",
  "author": "Your Name <your@email.com>",
  "publishConfig": {
    "access": "public",
    "provenance": true
  },
  "type": "module",
  "description": "Carbon Labs - Resize Bar",
  "license": "Apache-2.0",
  "repository": {
    "type": "git",
    "url": "https://github.com/carbon-design-system/carbon-labs",
    "directory": "src/components/Resizer"
  },
  "types": "./es/index.d.ts",
  "exports": {
    ".": {
      "import": "./es/index.js",
      "require": "./lib/index.js"
    },
    "./es/*": "./es/*",
    "./lib/*": "./lib/*",
    "./scss/*": "./scss/*"
  },
  "files": [
    "es",
    "lib",
    "scss"
  ],
  "scripts": {
    "build": "node ../../../tasks/build.js",
    "clean": "rm -rf {es,lib,scss}"
  },
  "devDependencies": {
    "@carbon-labs/utilities": "canary"
  }
}
