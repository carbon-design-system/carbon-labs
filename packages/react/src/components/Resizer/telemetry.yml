# yaml-language-server: $schema=https://unpkg.com/@ibm/telemetry-config-schema@v1/dist/config.schema.json

version: 1
projectId: c3e81c63-8688-4d9d-9f8f-2c0c369a780a
endpoint: https://www-api.ibm.com/ibm-telemetry/v1/metrics
collect:
  jsx:
    elements:
      allowedAttributeNames:
        # Resizer
        - children
        - className
        - onDoubleClick
        - onResize
        - onResizeEnd
        - orientation
        - thickness
        # React
        - key
        - ref
      allowedAttributeStringValues:
        # Resizer - orientation
        - horizontal
        - vertical
  npm:
    dependencies: null
  js:
    functions: {}
    tokens: null
