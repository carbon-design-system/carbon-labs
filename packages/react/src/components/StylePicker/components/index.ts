/**
 * @license
 *
 * Copyright IBM Corp. 2025
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

export { StylePickerColor } from './style-picker-color';
export { StylePickerIcon } from './style-picker-icon';
export { StylePickerGroup } from './style-picker-group';
export { StylePickerSection } from './style-picker-section';
export { StylePickerSections } from './style-picker-sections';
export { StylePickerOption } from './style-picker-option';
export { StylePicker } from './style-picker';
