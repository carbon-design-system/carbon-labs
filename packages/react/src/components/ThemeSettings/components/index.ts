/**
 * @license
 *
 * Copyright IBM Corp. 2025
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

export { ThemeSettings } from './ThemeSettings';
export { ThemeSwitcher } from './ThemeSwitcher';
export { ThemeSetDropdown } from './ThemeSetDropdown';
export { ThemeMenuComplement } from './ThemeMenuComplement';
export { themeSets } from './theme-settings-consts';
