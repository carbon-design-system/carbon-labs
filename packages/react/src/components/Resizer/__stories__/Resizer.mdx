import { ArgTypes, Canvas, Meta } from '@storybook/addon-docs/blocks';
import * as ResizerStories from './Resizer.stories';

<Meta isTemplate />

# Resizer

- **Initiative owner(s):** Nandan Devadula
- **Status:** Draft
- **Target library:** TBD
- **Target library maintainer(s) / PR Reviewer(s):** N/A
- **Support channel:** `#carbon-labs`

{/* <!-- START doctoc generated TOC please keep comment here to allow auto update --> */}
{/* <!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE --> */}

> ðŸ’¡ Check our
> [Stackblitz](https://stackblitz.com/github/carbon-design-system/carbon-labs/tree/main/examples/react/Resizer)
> example implementation.

[![Edit carbon-labs](https://developer.stackblitz.com/img/open_in_stackblitz.svg)](https://stackblitz.com/github/carbon-design-system/carbon-labs/tree/main/examples/react/Resizer)

## Table of Contents

- [Overview](#overview)
- [Getting started](#getting-started)

{/* <!-- END doctoc generated TOC please keep comment here to allow auto update --> */}

## Overview

`Resizer` is an atomic component extracted from `SplitPanel` and made into an
independent component, which can be reused wherever there is a resizing
requirement in the UI.

The `Resizer` captures the sibling sizes internally and resets on double-click.
This behavior can be overridden by using `onDoubleClick` to handle any custom
functionality.

The handle trigger cue can be enhanced if needed by writing a pseudo-class to
make the trigger easier to grab, or any custom elements can be passed as
children as seen in the `With Custom Handles` story. It can also be positioned
relative to the thickness prop to appear centered on the handle.

A gap can be added to the parent if the design demands.

The minimum inline size or block size of the panels can be customized based on
the design requirements.

For any kind of custom functionality, the component provides a callback
`onResize`, which gives access to `delta`, the difference in mouse or keyboard
movements relative to the initial position. along with the `event`.

Since most parts of this component are designed to be headless and
unopinionated, it can be extensively tweaked, providing opportunities for
various custom implementations.

<Canvas of={ResizerStories.TwoPanelsVertical} />

## A11y

The resizer is primarily a visual enhancement and may offer limited benefit for
screen-reader users, as the core UI elements remain accessible. However, for
improved screen reader support, developers can apply their own custom aria-label
to the resizer component ref available in `onResizeEnd` callback. This enables
users to receive live feedback on layout changes. A reference implementation of
this approach is included in the `Single Panel Bounded` example.

Note: `aria-valuenow`, `aria-valuemin`, and `aria-valuemax` must be set
manually, depending on how the resizer is used in the implementation.

## Getting started

Here's a quick example to get you started.

```bash
yarn add @carbon/react
yarn add @carbon-labs/react-resizer
```

### JS (via import)

```javascript
import { Resizer } from '@carbon-labs/react-resizer/es/index';

function App() {
  return <Resizer />;
}

export default App;
```

### SCSS

In your styles file import

```
@use '@carbon/react';
@use '@carbon-labs/react-resizer/scss/resizer';
```
