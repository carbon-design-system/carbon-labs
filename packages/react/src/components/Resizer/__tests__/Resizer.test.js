// cspell:ignore resizer
/* eslint-disable jsdoc/require-jsdoc, react/prop-types, react/forbid-dom-props */
/**
 * @license
 *
 * Copyright IBM Corp. 2025
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { render, fireEvent, screen } from '@testing-library/react';
import React from 'react';
import '@testing-library/jest-dom';
import { DEBOUNCE_DELAY } from '../components/Resizer';

import { Resizer } from '../components/Resizer';
jest.mock('./resize-bar.scss', () => ({}));

describe('Resizer', () => {
  let prev, next, resizer;
  const ResizableContainer = ({
    orientation = 'vertical',
    onResize,
    onResizeEnd,
    onDoubleClick,
    'data-testid': testId = 'resizer',
  }) => {
    const [prevSize, setPrevSize] = React.useState(100);
    const [nextSize, setNextSize] = React.useState(100);

    const handleResize = (_, delta) => {
      setPrevSize((p) => p + delta);
      setNextSize((n) => n - delta);
      onResize?.(_, delta);
    };

    const handleResizeEnd = (e, meta) => {
      onResizeEnd?.(e, meta);
    };

    const handleDoubleClick = () => {
      setPrevSize(100);
      setNextSize(100);
      onDoubleClick?.();
    };

    const styleKey = orientation === 'vertical' ? 'width' : 'height';

    return (
      <div
        style={{
          display: 'flex',
          flexDirection: orientation === 'vertical' ? 'row' : 'column',
        }}>
        <div data-testid="prev" style={{ [styleKey]: `${prevSize}px` }} />
        <Resizer
          orientation={orientation}
          data-testid={testId}
          onResize={handleResize}
          onResizeEnd={handleResizeEnd}
          onDoubleClick={handleDoubleClick}
        />
        <div data-testid="next" style={{ [styleKey]: `${nextSize}px` }} />
      </div>
    );
  };

  const setupDOM = (orientation = 'vertical', props = {}) => {
    const utils = render(
      <ResizableContainer orientation={orientation} {...props} />
    );
    prev = utils.getByTestId('prev');
    next = utils.getByTestId('next');
    resizer = utils.getByTestId(props['data-testid'] || 'resizer');
    return utils;
  };

  describe('renders as expected - Component API', () => {
    it('should match snapshot', () => {
      const { container } = render(
        <Resizer orientation="vertical" data-testid="resizer" />
      );
      expect(container).toMatchSnapshot();
    });

    it('should set correct aria attributes', () => {
      const { getByRole } = render(
        <Resizer orientation="horizontal" data-testid="resizer" />
      );
      const el = getByRole('separator');
      expect(el).toHaveAttribute('aria-orientation', 'horizontal');
      expect(el).toHaveAttribute('aria-live', 'assertive');
      expect(el).toHaveAttribute('tabindex', '0');
    });

    it('should apply custom className', () => {
      const { getByRole } = render(
        <Resizer orientation="vertical" className="custom-class" />
      );
      expect(getByRole('separator')).toHaveClass('custom-class');
    });
  });

  describe('mouse events', () => {
    it('calls onResize and onResizeEnd on mouse drag', () => {
      const onResize = jest.fn();
      const onResizeEnd = jest.fn();
      setupDOM('vertical', { onResize, onResizeEnd });

      fireEvent.mouseDown(resizer, { clientX: 50 });
      fireEvent.mouseMove(window, { clientX: 70 });
      fireEvent.mouseUp(window, { clientX: 70 });

      expect(onResize).toHaveBeenCalledWith(expect.any(Object), 20);
      expect(onResizeEnd).toHaveBeenCalledWith(
        expect.any(Object),
        expect.any(Object)
      );
    });

    it('removes event listeners on unmount', () => {
      const { unmount, getByRole } = render(
        <Resizer orientation="vertical" data-testid="resizer" />
      );
      const el = getByRole('separator');
      fireEvent.mouseDown(el, { clientX: 0 });
      unmount();
      fireEvent.mouseMove(window, { clientX: 10 });
      fireEvent.mouseUp(window, { clientX: 10 });
    });
  });

  describe('keyboard events', () => {
    it('handles ArrowLeft/ArrowRight for vertical', () => {
      setupDOM('vertical');
      fireEvent.keyDown(resizer, { key: 'ArrowRight' });
      expect(prev).toHaveStyle('width: 105px');
      expect(next).toHaveStyle('width: 95px');

      fireEvent.keyDown(resizer, { key: 'ArrowLeft' });
      expect(prev).toHaveStyle('width: 100px');
      expect(next).toHaveStyle('width: 100px');
    });

    it('handles shift+ArrowRight for larger step', () => {
      setupDOM('vertical');
      fireEvent.keyDown(resizer, { key: 'ArrowRight', shiftKey: true });
      expect(prev).toHaveStyle('width: 125px');
      expect(next).toHaveStyle('width: 75px');
    });

    // it('handles Home/End keys', () => {
    //   setupDOM('vertical');
    //   fireEvent.keyDown(resizer, { key: 'End' });
    //   expect(prev).toHaveStyle('width: 200px');
    //   expect(next).toHaveStyle('width: 0px');

    //   fireEvent.keyDown(resizer, { key: 'Home' });
    //   expect(prev).toHaveStyle('width: 0px');
    //   expect(next).toHaveStyle('width: 200px');
    // });

    it('ignores unrelated keys', () => {
      setupDOM('vertical');
      fireEvent.keyDown(resizer, { key: 'a' });
      expect(prev).toHaveStyle('width: 100px');
      expect(next).toHaveStyle('width: 100px');
    });

    it('calls onResize and onResizeEnd with keyboard', async () => {
      const onResize = jest.fn();
      const onResizeEnd = jest.fn();

      setupDOM('vertical', {
        onResize,
        onResizeEnd,
        'data-testid': 'resizer2',
      });

      const resizer2 = screen.getByTestId('resizer2');
      fireEvent.keyDown(resizer2, { key: 'ArrowRight' });

      expect(onResize).toHaveBeenCalledWith(expect.any(Object), 5);
      await new Promise((resolve) => setTimeout(resolve, DEBOUNCE_DELAY));
      expect(onResizeEnd).toHaveBeenCalled();
    });
  });

  describe('double click', () => {
    it('resets sizes on double click', () => {
      setupDOM('vertical');
      fireEvent.keyDown(resizer, { key: 'ArrowRight', shiftKey: true }); // simulate resize
      fireEvent.doubleClick(resizer);
      expect(prev).toHaveStyle('width: 100px');
      expect(next).toHaveStyle('width: 100px');
    });

    it('calls onDoubleClick if provided', () => {
      const onDoubleClick = jest.fn();
      const { getByTestId } = render(
        <ResizableContainer
          orientation="vertical"
          onDoubleClick={onDoubleClick}
          data-testid="resizer2"
        />
      );
      const resizer2 = getByTestId('resizer2');
      fireEvent.doubleClick(resizer2);
      expect(onDoubleClick).toHaveBeenCalled();
    });
  });

  describe('ref forwarding', () => {
    it('forwards ref if provided', () => {
      const ref = React.createRef();
      render(
        <Resizer orientation="vertical" ref={ref} data-testid="resizer" />
      );
      expect(ref.current).toBeInstanceOf(HTMLElement);
    });
  });
});
