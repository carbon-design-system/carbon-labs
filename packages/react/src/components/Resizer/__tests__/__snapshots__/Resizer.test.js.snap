// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Resizer renders as expected - Component API should match snapshot 1`] = `
<div>
  <div
    aria-live="assertive"
    aria-orientation="vertical"
    class="clabs__resizer clabs__resizer--vertical"
    data-testid="resizer"
    role="separator"
    style="inline-size: 0.25rem;"
    tabindex="0"
  >
    <span
      class="sr-only"
    >
      Use arrow keys to resize, hold Shift for larger steps. Double-click to reset.
    </span>
  </div>
</div>
`;
